lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -0,0 +1,485 @@
+        {
+            goto LReturn;
+        }
+
+        toLangTagResultLength = uloc_toLanguageTag(icuLocaleId, icuLangTag, ULOC_FULLNAME_CAPACITY, TRUE, &error);
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
+            }
+            else
+            {
+                AssertMsg(false, "uloc_toLanguageTag: unexpected error (besides U_ILLEGAL_ARGUMENT_ERROR)");
+            }
+
+            goto LReturn;
+        }
+
+    LReturn:
+        return success;
+    }
+
+    HRESULT NormalizeLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch,
+        _Out_ char16 *normalized, _Out_ size_t *normalizedLength)
+    {
+        bool success = false;
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+        char icuLocaleId[ULOC_FULLNAME_CAPACITY] = { 0 };
+        char icuLangTag[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t parsedLength = 0;
+        int32_t forLangTagResultLength = 0;
+        int32_t toLangTagResultLength = 0;
+
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        unsigned char *inputLangTagUtf8 = new unsigned char[inputLangTagUtf8SizeAllocated];
+        StringBufferAutoPtr<unsigned char> guard(inputLangTagUtf8);
+
+        if (!inputLangTagUtf8)
+        {
+            AssertOrFailFastMsg(false, "OOM: failed to allocate inputLangTagUtf8.");
+            goto LReturn;
+        }
+
+        // explicitly zero the allocated array
+        for (size_t index = 0; index < inputLangTagUtf8SizeAllocated; ++index)
+        {
+            inputLangTagUtf8[index] = 0;
+        }
+
+        utf8::EncodeIntoAndNullTerminate(inputLangTagUtf8, languageTag, cch);
+
+        // Convert input language tag to a locale ID for use in uloc_toLanguageTag API.
+        forLangTagResultLength = uloc_forLanguageTag(reinterpret_cast<const char *>(inputLangTagUtf8), icuLocaleId, ULOC_FULLNAME_CAPACITY, &parsedLength, &error);
+        success = forLangTagResultLength && parsedLength && U_SUCCESS(error);
+        if (!success)
+        {
+            AssertMsg(false, "uloc_forLanguageTag failed");
+            goto LReturn;
+        }
+
+        // Try to convert icuLocaleId (locale ID version of input locale string) to BCP47 language tag, using strict checks
+        toLangTagResultLength = uloc_toLanguageTag(icuLocaleId, icuLangTag, ULOC_FULLNAME_CAPACITY, TRUE, &error);
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
```

Comment:
Are these really errors we cannot recover from? Are the arguments provided by user code? If not, I suppose this is probably fine. We might want to consider returning a more specific error code instead of just `E_INVALIDARG`, though.

Reply:
For these specific errors, I could see us not wanting to assert in the case of an illegal argument, which may simply mean that the input tag couldn't be normalized. That's an expected error and probably shouldn't stop a debug build from running. With that being said, the only "recovery" would be to propagate the error back to the user and let them decide what to do with an un-normalizable langtag.

Reply:
Also, the logic should already have taken care of that situation in a spec-complaint way. This Assert is because that sort of issue shouldn't get to this point.